packages/astro/src/i18n/middleware.ts

```diff
@@ -54,30 +55,52 @@ export function createI18nMiddleware(
 
 		if (response instanceof Response) {
 			const pathnameContainsDefaultLocale = url.pathname.includes(`/${defaultLocale}`);
-			if (i18n.routing === 'prefix-other-locales' && pathnameContainsDefaultLocale) {
-				const newLocation = url.pathname.replace(`/${defaultLocale}`, '');
-				response.headers.set('Location', newLocation);
-				return new Response(null, {
-					status: 404,
-					headers: response.headers,
-				});
-			} else if (i18n.routing === 'prefix-always') {
-				if (url.pathname === base + '/' || url.pathname === base) {
-					if (trailingSlash === 'always') {
-						return context.redirect(`${appendForwardSlash(joinPaths(base, i18n.defaultLocale))}`);
-					} else {
-						return context.redirect(`${joinPaths(base, i18n.defaultLocale)}`);
+			switch (i18n.routing) {
+				case 'pathname-prefix-other-locales': {
+					if (pathnameContainsDefaultLocale) {
+						const newLocation = url.pathname.replace(`/${defaultLocale}`, '');
+						response.headers.set('Location', newLocation);
+						return new Response(null, {
+							status: 404,
+							headers: response.headers,
+						});
+					}
+					break;
+				}
+
+				case 'pathname-prefix-always-no-redirect': {
+					// We return a 404 if:
+					// - the current path isn't a root. e.g. / or /<base>
+					// - the URL doesn't contain a locale
+					const isRoot = url.pathname.startsWith(base) || url.pathname === '/';
+					if (!(isRoot || pathnameHasLocale(url.pathname, i18n.locales))) {
+						return new Response(null, {
+							status: 404,
+							headers: response.headers,
+						});
 					}
+					break;
 				}
 
-				// Astro can't know where the default locale is supposed to be, so it returns a 404 with no content.
-				else if (!pathnameHasLocale(url.pathname, i18n.locales)) {
-					return new Response(null, {
-						status: 404,
-						headers: response.headers,
-					});
+				case 'pathname-prefix-always': {
+					if (url.pathname === base + '/' || url.pathname === base) {
+						if (trailingSlash === 'always') {
```

Comment:
Unrelated: It looks like the prefix is only added when `trailingSlash` is set to `always`. Shouldn't the prefix also be added for other `trailingSlash` configurations?

Reply:
Right, although to check the other variants, we need `build.format`, which is a piece of information that we don't have in SSR (in the SSR manifest), that's why I only check `always`. Should store `build.format` in the SSRManifest?

Reply:
Ah ok that makes sense. I think it's fine to store `build.format` then if it helps improve things here. We can definitely improve this in a later PR.